phaser/pipeline.py

```diff
@@ -227,6 +240,11 @@ def save_extra_outputs(self):
                 save_csv(filename, item.data)
                 logger.info(f"Extra output {item.name} saved to {self.working_dir}")
                 item.to_save = False
+                # If the output is meant to be used as a source also, then set
+                # the data into the context as a source so that the phase that
+                # needs it can access the data.
+                if item.name in self.extra_sources:
+                    self.context.set_source(item.name, item.data)
```

Comment:
Might it make sense for both inputs and outputs to be in the same indexed list of things?
* it would be good to avoid any overwriting of names - and that would be more inherent in this data model
* no need to copy output-> source
 
I guess related to that is the decision whether we allow a Phase access to any RWO already on the context, or only the RWOs that are declared for it  to use in the construction of the Pipeline

Reply:
Agreed on making sources and outputs be the same collection of things.

About offering to a Phase only the RWO that it declares it creates or accesses, that would be a nice additional safety mechanism.  I think to do that best, we would want to pass in a context to the phase in the `run()` function so that the pipeline can offer "views" of the context that only have in them the correct sources.  I'm starting to have delusions of needing some functional data structures.

Reply:
Speaking of functional data structures, merging the sources and outputs into a single structure also makes me want to offer the sources as immutable data structures to the steps so that they are not inadvertently changed.  Holding myself back from that for the moment.